_posts/2018-10-12-kafka-stream-introduction.md

```diff
@@ -0,0 +1,257 @@
+These represent a period of activity separated by a defined gap of inactivity.
+All events within that gap will be merged with an existing session.
+If the gap is too large a new session window will be created, the size of the window itself will thus vary.
+<img class="image fit" style="margin:0px auto; max-width: 500px;" alt="Session Window" src="/img/kafka/streams-session-windows-02.png">
+
+## The Practical Part
+### Disclaimer
+This project is intended as a first step into the world of streaming, so some shortcuts were taken, and not all design decisions are production ready. 
+A good example is the use of strings as the content of the messages, this should be done in a more structured way (with [Avro](https://avro.apache.org/) for example).
+
+### Setup of the project
+This is the really easy part, to use the streaming api from Kafka only 1 dependency must be added. 
+Here is the example to do it in Maven. 
+```xml
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.kafka</groupId>
+            <artifactId>kafka-streams</artifactId>
+            <version>1.1.0</version>
+        </dependency>
+    </dependencies>
+```
+And that is it. Sometimes live can be simple :)
+
+### Creation of the input
+In the real world this would be done by the IOT devices that send there data through the network to the central system. 
+But since it is not easy for demo purposes to have a sensor and a door nearby, and even less handy to open and close it a couple of hundred times to test it out,
+I created a simulator that just sends data to the Kafka cluster.
+
+This simulator creates 2 kinds of messages:
+`key = 0E7E346406100585, value = T_7`
+configuration information about on which floor a certain device is located.
+`key = 0E7E346406100585, value = pulse@1496309915`
+each time a person opens the door the key is the unique id of the device, and then a pulse and the time at which it occurred
+
+```java
+     List<String> devices = new ArrayList<>();
+    devices.add(UUID.randomUUID().toString());
+    devices.add(UUID.randomUUID().toString());  
+    devices.add(UUID.randomUUID().toString());
+
+    KafkaProducer producer = new KafkaProducer(props);
+
+    //send the device information
+    for (int i = 0; i < devices.size(); i++) {
+        String val = String.format("%s@%s@%s", "T", "" + (i + 1), System.currentTimeMillis());
+        producer.send(new ProducerRecord("stream_in_dev", devices.get(i), val));
+    }
+
+    try {
+        Thread.sleep(1000);
+    } catch (InterruptedException e) {
+    }
+
+    new Random().ints(10, 0, devices.size()).forEach(i -> {
+        producer.send(new ProducerRecord("stream_in", devices.get(i), "pulse@" + System.currentTimeMillis()));
+        try {
+            Thread.sleep(250);
+        } catch (InterruptedException e) {}
+    });
+
+    producer.close();
+```
+
+This wil create 3 floors with a random device id, and afterwards send 10 times an event for a random door that it is opened.
```

Comment:
```suggestion
This wil create three floors with a random device id, and afterwards it will send an event for a random door being opened 10 times.
```